core/config/config_helper.hpp

```diff
@@ -0,0 +1,220 @@
+        static std::string str() { return _str; } \
+    }
+
+TYPE_STRING_OVERLOAD(void, "void");
+TYPE_STRING_OVERLOAD(double, "double");
+TYPE_STRING_OVERLOAD(float, "float");
+TYPE_STRING_OVERLOAD(std::complex<double>, "complex<double>");
+TYPE_STRING_OVERLOAD(std::complex<float>, "complex<float>");
+TYPE_STRING_OVERLOAD(int32, "int");
+TYPE_STRING_OVERLOAD(int64, "int64");
+
+
+template <typename T>
+inline std::shared_ptr<T> get_pointer(const pnode& config,
+                                      const registry& context,
+                                      type_descriptor td)
+{
+    std::shared_ptr<T> ptr;
+    using T_non_const = std::remove_const_t<T>;
+    ptr = context.search_data<T_non_const>(config.get_string());
+    assert(ptr.get() != nullptr);
+    return std::move(ptr);
+}
+
+
+template <typename T>
+inline std::vector<deferred_factory_parameter<T>> get_factory_vector(
+    const pnode& config, const registry& context, type_descriptor td)
+{
+    std::vector<deferred_factory_parameter<T>> res;
+    // for loop in config
+    if (config.get_tag() == pnode::tag_t::array) {
+        for (const auto& it : config.get_array()) {
+            res.push_back(get_factory<T>(it, context, td));
+        }
+    } else {
+        // only one config can be passed without array
+        res.push_back(get_factory<T>(config, context, td));
+    }
+
+    return res;
+}
+
+
+template <typename IndexType>
+inline
+    typename std::enable_if<std::is_integral<IndexType>::value, IndexType>::type
+    get_value(const pnode& config)
+{
+    auto val = config.get_integer();
+    GKO_THROW_IF_INVALID(
+        val <= std::numeric_limits<IndexType>::max() &&
+            val >= std::numeric_limits<IndexType>::min(),
+        "the config value is out of the range of the require type.");
+    return static_cast<IndexType>(val);
+}
+
+template <typename ValueType>
+inline typename std::enable_if<std::is_floating_point<ValueType>::value,
+                               ValueType>::type
+get_value(const pnode& config)
+{
+    auto val = config.get_real();
+    // the max, min of floating point only consider positive value.
+    GKO_THROW_IF_INVALID(
+        val <= std::numeric_limits<ValueType>::max() &&
+            val >= -std::numeric_limits<ValueType>::max(),
+        "the config value is out of the range of the require type.");
+    return static_cast<ValueType>(val);
+}
+
+template <typename ValueType>
+inline typename std::enable_if<gko::is_complex_s<ValueType>::value,
+                               ValueType>::type
+get_value(const pnode& config)
+{
+    using real_type = gko::remove_complex<ValueType>;
+    if (config.get_tag() == pnode::tag_t::real) {
+        return static_cast<ValueType>(get_value<real_type>(config));
+    } else if (config.get_tag() == pnode::tag_t::array) {
+        return ValueType{get_value<real_type>(config.get(0)),
+                         get_value<real_type>(config.get(1))};
```

Comment:
I would suggest using
```
real: value,
imag: value
```
for complex. Otherwise it might not be clear from the config that a complex value is described.

Reply:
I guess this is important for #1289 ? Otherwise we don't really need to specify any complex numbers explicitly yet, I believe

Reply:
I have mentioned the style in the top of config.hpp
it is also related to the weighted IR

Reply:
I'm not sure if Richardson iterations make sense for values outside the real numbers, since it's usually intended as an interpolation between old and new value compared to a regular IR. But otherwise I agree.